File: Dev/Swift/Sources/SKAISwiftUI/PlaceholderScene.swift
import SwiftUI
#if canImport(UIKit)
import UIKit
#elseif canImport(AppKit)
import AppKit
#endif

/// The placeholder body for a UI Map scene. A scaffolded scene view's `body`
/// invokes this.
///
/// Root, pushed, modal, child-hosted, and tab scenes render full-bleed. A scene
/// embedded as composite content draws an inset fill so the nesting remains
/// visible. Layout is inferred from the declared routes, tabs, and content.
///
/// Shows the scene's name as a visible label, an optional "Routes" menu, and a
/// dismiss "✕" driven by `@Environment(\.dismiss)`.
public struct PlaceholderScene<EmbeddedContent: View>: View {

    // MARK: - API

    /// Scene with embedded composite/child content — optionally also tabbed.
    /// - Parameters:
    ///   - title: The scene's name (the visible identifier).
    ///   - routes: Outgoing nav/modal routes for the "Routes" menu; empty hides the menu.
    ///   - tabs: Tab children rendered as a real `TabView`; empty for a non-tabbed scene.
    ///   - embedded: Composite/child scene placeholders, instantiated directly; each is
    ///     itself a `PlaceholderScene` and carries its own border.
    public init(
        title: String,
        domainColor: Color = .accentColor,
        routes: [PlaceholderRoute] = [],
        tabs: [PlaceholderTab] = [],
        @ViewBuilder embedded: () -> EmbeddedContent
    ) {
        self.title = title
        self.domainColor = domainColor
        self.routes = routes
        self.tabs = tabs
        self.embedded = embedded()
        self.hasEmbedded = true
    }

    /// Leaf scene (no embedded content) or a tabs-only container.
    public init(
        title: String,
        domainColor: Color = .accentColor,
        routes: [PlaceholderRoute] = [],
        tabs: [PlaceholderTab] = []
    ) where EmbeddedContent == EmptyView {
        self.title = title
        self.domainColor = domainColor
        self.routes = routes
        self.tabs = tabs
        self.embedded = EmptyView()
        self.hasEmbedded = false
    }

    // MARK: - Constants

    private static var insetPadding: CGFloat { 6 }
    private static var contentPadding: CGFloat { 12 }
    private static var cornerRadius: CGFloat { 10 }

    // MARK: - Variables

    private let title: String
    private let domainColor: Color
    private let routes: [PlaceholderRoute]
    private let tabs: [PlaceholderTab]
    private let embedded: EmbeddedContent
    private let hasEmbedded: Bool

    @Environment(\.dismiss) private var dismiss
    @Environment(\.placeholderCrumbs) private var crumbs
    @Environment(\.placeholderIsEmbedded) private var isEmbedded
    @Environment(\.placeholderShowsDismiss) private var showsDismiss

    @State private var breadcrumbScroll = ScrollPosition(edge: .trailing)

    // MARK: - Lifecycle

    public var body: some View {
        layout
            .background {
                if !isFullBleed {
                    RoundedRectangle(cornerRadius: Self.cornerRadius)
                        .fill(Colors.composite)
                }
            }
            .padding(isFullBleed ? 0 : Self.insetPadding)
            .modifier(InlineNavTitle(title: title, enabled: isFullBleed))
            .tint(domainColor)
    }

    // MARK: - Layout

    /// A full-bleed scene fills edge-to-edge with no box border or inset gap.
    /// Composite children are inset; presentation boundaries and destinations
    /// are full-bleed. Breadcrumbs/dismiss context takes precedence when a scene
    /// nested in a composite presents another destination.
    private var isFullBleed: Bool { !isEmbedded || !crumbs.isEmpty || showsDismiss }

    /// This scene's inherited crumbs plus itself — handed down to tab content so
    /// the breadcrumb accumulates through the tab boundary.
    private var effectiveCrumbs: [PlaceholderCrumb] {
        crumbs + [PlaceholderCrumb(title: title, color: domainColor, routes: routes)]
    }

    @ViewBuilder
    private var layout: some View {
        if !tabs.isEmpty {
            // Tabbed: a full-bleed TabView owns the screen edges (tab bar at the
            // bottom). The breadcrumb is delegated to the tab content, which
            // accumulates this scene into its crumbs and renders it under its
            // own nav bar.
            tabView
        } else {
            VStack(alignment: .leading, spacing: 16) {
                breadcrumb
                if hasEmbedded {
                    embedded
                        .environment(\.placeholderCrumbs, [])
                        .environment(\.placeholderIsEmbedded, true)
                        .environment(\.placeholderShowsDismiss, false)
                }
            }
            .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
            .padding(Self.contentPadding)
        }
    }

    private var breadcrumb: some View {
        HStack(spacing: 8) {
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 6) {
                    ForEach(crumbs) { crumb in
                        titleMenu(title: crumb.title, routes: crumb.routes, color: crumb.color)
                        Text("›").font(.headline).foregroundStyle(.secondary)
                    }
                    titleMenu(title: title, routes: routes, color: domainColor)
                }
            }
            .scrollPosition($breadcrumbScroll)

            if showsDismiss {
                dismissButton
            }
        }
    }

    @ViewBuilder
    private func titleMenu(title: String, routes: [PlaceholderRoute], color: Color) -> some View {
        let visible = routes.filter { !$0.isCurrent }
        let pill = Text(title)
            .font(.subheadline)
            .foregroundStyle(.black)
            .padding(.vertical, 3)
            .padding(.horizontal, 8)
            .background(RoundedRectangle(cornerRadius: 8).fill(color))
        if visible.isEmpty {
            pill.opacity(0.5)
        } else {
            Menu {
                ForEach(PlaceholderRouteKind.allCases, id: \.self) { kind in
                    let kindRoutes = visible.filter { $0.kind == kind }
                    if !kindRoutes.isEmpty {
                        Section(kind.title) {
                            ForEach(kindRoutes) { route in
                                Button(route.label) { route.action() }
                            }
                        }
                    }
                }
            } label: {
                pill
            }
        }
    }

    private var tabView: some View {
        TabView {
            ForEach(tabs) { tab in
                tab.content
                    .environment(\.placeholderCrumbs, effectiveCrumbs)
                    .environment(\.placeholderShowsDismiss, false)
                    .tabItem { Label(tab.label, systemImage: resolvedSymbol(tab.systemImage)) }
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
    }

    /// Returns `name` if it's a real SF Symbol, else the default — guards against
    /// a scaffold picking a symbol that doesn't exist (which would render blank).
    private func resolvedSymbol(_ name: String) -> String {
        #if canImport(UIKit)
        return UIImage(systemName: name) != nil ? name : "square"
        #elseif canImport(AppKit)
        return NSImage(systemSymbolName: name, accessibilityDescription: nil) != nil ? name : "square"
        #else
        return name
        #endif
    }

    private var dismissButton: some View {
        Button {
            dismiss()
        } label: {
            Image(systemName: "xmark")
                .font(.caption.weight(.bold))
                .foregroundStyle(.black)
                .padding(7)
                .background(Circle().fill(domainColor))
        }
        .buttonStyle(.plain)
        .accessibilityLabel("Dismiss")
    }
}

private struct PlaceholderIsEmbeddedKey: EnvironmentKey {
    static let defaultValue = false
}

private extension EnvironmentValues {
    var placeholderIsEmbedded: Bool {
        get { self[PlaceholderIsEmbeddedKey.self] }
        set { self[PlaceholderIsEmbeddedKey.self] = newValue }
    }
}

// MARK: - Nav title

/// Sets the scene's nav-bar title (inline on iOS). A no-op outside a
/// `NavigationStack`. `navigationBarTitleDisplayMode` is iOS-only, so it's
/// guarded for the package's macOS build.
private struct InlineNavTitle: ViewModifier {
    let title: String
    let enabled: Bool

    @ViewBuilder
    func body(content: Content) -> some View {
        if enabled {
            #if os(iOS)
            content
                .navigationTitle(title)
                .navigationBarTitleDisplayMode(.inline)
            #else
            content.navigationTitle(title)
            #endif
        } else {
            content
        }
    }
}

// MARK: - Previews

#Preview("Leaf") {
    PlaceholderScene(title: "Cancel Appointment")
}

#Preview("Leaf + routes") {
    PlaceholderScene(
        title: "Profile",
        routes: [
            PlaceholderRoute(label: "Debug", kind: .nav) {},
            PlaceholderRoute(label: "Web", kind: .nav) {},
        ]
    )
}

#Preview("Container") {
    PlaceholderScene(
        title: "Appointment Details",
        routes: [PlaceholderRoute(label: "Cancel Appointment", kind: .sheet) {}]
    ) {
        PlaceholderScene(title: "Thumbnail Slider")
        PlaceholderScene(title: "Record Warning")
    }
}

#Preview("Tabbed") {
    PlaceholderScene(
        title: "Main",
        tabs: [
            PlaceholderTab(label: "Home") { PlaceholderScene(title: "Home") },
            PlaceholderTab(label: "Library") { PlaceholderScene(title: "Library") },
        ]
    )
}

#Preview("Tabbed + embedded") {
    PlaceholderScene(
        title: "Pipeline",
        routes: [PlaceholderRoute(label: "Menu", kind: .nav) {}],
        tabs: [
            PlaceholderTab(label: "Appointments") { PlaceholderScene(title: "Appointments Stage") },
            PlaceholderTab(label: "Closed") { PlaceholderScene(title: "Closed Pipeline Stage") },
        ]
    ) {
        PlaceholderScene(title: "Header Banner")
    }
}

File: Dev/Compose/skai-compose-android/src/main/java/com/swiftkickmobile/skai/compose/placeholder/PlaceholderScene.kt
package com.swiftkickmobile.skai.compose.placeholder

import androidx.compose.animation.EnterTransition
import androidx.compose.animation.ExitTransition
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Close
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.navigation.NavDestination.Companion.hasRoute
import androidx.navigation.compose.NavHost
import androidx.navigation.compose.composable
import androidx.navigation.compose.currentBackStackEntryAsState
import androidx.navigation.compose.rememberNavController
import com.swiftkickmobile.skai.compose.navigation.navigateToTab

/** Shared layout constants mirroring SKAISwiftUI's `PlaceholderScene`. */
private object PlaceholderDimens {
    val inset = 6.dp
    val content = 12.dp
    val corner = 10.dp
}

/**
 * The placeholder body for a UI Map scene. A scaffolded scene composable invokes
 * this. Mirrors SKAISwiftUI's `PlaceholderScene`.
 *
 * Embedded composite scenes draw their own inset gray fill so nesting remains
 * visible. Layout is inferred from what's declared and presentation context:
 * - Root / leaf: full-bleed with a breadcrumb row.
 * - Full-bleed (carries crumbs from a child host / tab / push, or is modal):
 *   fills edge-to-edge, breadcrumb accumulates ancestors.
 * - Tabbed: a real bottom-tab scaffold over a tab `NavHost` switched with
 *   `navigateToTab`; tab content accumulates this scene.
 *
 * @param title The scene's name (the visible identifier).
 * @param domainColor The scene's domain tint.
 * @param routes Outgoing routes for this scene's breadcrumb menu; empty hides it.
 * @param tabs Tab children rendered as a bottom-tab scaffold, each a typed
 *   destination in the scene's tab `NavHost`; empty for non-tabbed.
 * @param embedded Composite/child scene placeholders, each itself a
 *   `PlaceholderScene` carrying its own inset fill; empty for a leaf scene.
 */
@Composable
fun PlaceholderScene(
    title: String,
    domainColor: Color = PlaceholderColors.domain(0),
    routes: List<PlaceholderRoute> = emptyList(),
    tabs: List<PlaceholderTab> = emptyList(),
    embedded: List<@Composable () -> Unit> = emptyList(),
) {
    val crumbs = LocalPlaceholderCrumbs.current
    val isEmbedded = LocalPlaceholderIsEmbedded.current
    val showsDismiss = LocalPlaceholderShowsDismiss.current

    // Roots and presentation destinations are full-bleed. Only a scene invoked
    // directly as embedded composite content keeps the inset gray treatment;
    // presentation context takes precedence if that scene routes onward.
    val isFullBleed = !isEmbedded || crumbs.isNotEmpty() || showsDismiss

    when {
        tabs.isNotEmpty() -> TabScaffold(title, domainColor, routes, tabs, crumbs)
        isFullBleed -> FullBleedScene(title, domainColor, routes, crumbs, showsDismiss, embedded)
        else -> EmbeddedScene(title, domainColor, routes, crumbs, showsDismiss, embedded)
    }
}

// MARK: - Layout

/**
 * A full-bleed scene (child, pushed, or modal). Gets a top app bar carrying the
 * title (and a back arrow on pushed scenes) — the Compose stand-in for SwiftUI's
 * `NavigationStack` nav bar, which also pushes content clear of the status bar.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun FullBleedScene(
    title: String,
    domainColor: Color,
    routes: List<PlaceholderRoute>,
    crumbs: List<PlaceholderCrumb>,
    showsDismiss: Boolean,
    embedded: List<@Composable () -> Unit>,
) {
    val onBack = LocalPlaceholderBack.current
    Column(Modifier.fillMaxSize()) {
        CenterAlignedTopAppBar(
            title = { Text(title) },
            navigationIcon = {
                if (onBack != null) {
                    IconButton(onClick = onBack) {
                        Icon(Icons.AutoMirrored.Filled.ArrowBack, contentDescription = "Back")
                    }
                }
            },
        )
        SceneContent(
            title, domainColor, routes, crumbs, showsDismiss, embedded,
            Modifier.fillMaxSize(),
        )
    }
}

/** An embedded composite child — an inset, gray-filled box with no app bar. */
@Composable
private fun EmbeddedScene(
    title: String,
    domainColor: Color,
    routes: List<PlaceholderRoute>,
    crumbs: List<PlaceholderCrumb>,
    showsDismiss: Boolean,
    embedded: List<@Composable () -> Unit>,
) {
    SceneContent(
        title, domainColor, routes, crumbs, showsDismiss, embedded,
        Modifier
            .fillMaxSize()
            .padding(PlaceholderDimens.inset)
            .clip(RoundedCornerShape(PlaceholderDimens.corner))
            .background(PlaceholderColors.composite),
    )
}

/** Shared breadcrumb + body, used by both the full-bleed and embedded layouts. */
@Composable
private fun SceneContent(
    title: String,
    domainColor: Color,
    routes: List<PlaceholderRoute>,
    crumbs: List<PlaceholderCrumb>,
    showsDismiss: Boolean,
    embedded: List<@Composable () -> Unit>,
    modifier: Modifier,
) {
    Column(
        modifier = modifier.padding(PlaceholderDimens.content),
        verticalArrangement = Arrangement.spacedBy(16.dp),
    ) {
        Breadcrumb(title, domainColor, routes, crumbs, showsDismiss)
        // Embedded composites are greedy: each takes an equal share of the
        // remaining height (mirrors SwiftUI's maxHeight-infinity children). Their
        // nesting is shown by the box, so the breadcrumb chain resets.
        embedded.forEach { child ->
            Box(
                modifier = Modifier
                    .weight(1f)
                    .fillMaxWidth(),
            ) {
                CompositionLocalProvider(
                    LocalPlaceholderCrumbs provides emptyList(),
                    LocalPlaceholderIsEmbedded provides true,
                    LocalPlaceholderShowsDismiss provides false,
                ) {
                    child()
                }
            }
        }
    }
}

/**
 * Hosts a child-routed scene full-bleed.
 *
 * A child route replaces the parent's content entirely, so the parent draws no
 * box of its own. Instead it renders the selected child filling the available
 * space and appends its identity (title + child-switcher routes) to the
 * breadcrumb the child shows, so the parent stays reachable.
 */
@Composable
fun PlaceholderChildHost(
    title: String,
    domainColor: Color,
    routes: List<PlaceholderRoute> = emptyList(),
    content: @Composable () -> Unit,
) {
    val crumbs = LocalPlaceholderCrumbs.current
    CompositionLocalProvider(
        LocalPlaceholderCrumbs provides crumbs + PlaceholderCrumb(title, domainColor, routes),
    ) {
        Box(Modifier.fillMaxSize()) { content() }
    }
}

// MARK: - Tabs

/**
 * A bottom-tab scaffold over the scene's tab `NavHost`. The first tab is the
 * primary tab (start destination); switching goes through [navigateToTab], so
 * each tab keeps its own nested state and back from any tab's root returns to
 * the primary tab.
 */
@Composable
private fun TabScaffold(
    title: String,
    domainColor: Color,
    routes: List<PlaceholderRoute>,
    tabs: List<PlaceholderTab>,
    parentCrumbs: List<PlaceholderCrumb>,
) {
    val tabNavController = rememberNavController()
    val currentDestination = tabNavController.currentBackStackEntryAsState().value?.destination

    // Delegate the breadcrumb to the tab content: accumulate this scene into the
    // crumbs the selected tab renders under its own layout.
    val effectiveCrumbs = parentCrumbs + PlaceholderCrumb(title, domainColor, routes)

    Scaffold(
        bottomBar = {
            NavigationBar {
                tabs.forEach { tab ->
                    NavigationBarItem(
                        selected = currentDestination?.hasRoute(tab.route::class) == true,
                        onClick = { tabNavController.navigateToTab(tab.route) },
                        icon = { Icon(tab.icon, contentDescription = tab.label) },
                        label = {
                            Text(
                                text = tab.label,
                                maxLines = 1,
                                softWrap = false,
                                overflow = TextOverflow.Ellipsis,
                            )
                        },
                    )
                }
            }
        },
    ) { insets ->
        // Consume the insets so each tab's content (a full-bleed scene with its
        // own top app bar) doesn't re-apply the status-bar inset on top of these.
        Box(
            Modifier
                .padding(insets)
                .consumeWindowInsets(insets)
                .fillMaxSize(),
        ) {
            CompositionLocalProvider(
                LocalPlaceholderCrumbs provides effectiveCrumbs,
                LocalPlaceholderIsEmbedded provides false,
            ) {
                // Tabs swap in place — no push animation.
                NavHost(
                    navController = tabNavController,
                    startDestination = tabs.first().route,
                    enterTransition = { EnterTransition.None },
                    exitTransition = { ExitTransition.None },
                ) {
                    tabs.forEach { tab ->
                        composable(route = tab.route::class) { tab.content() }
                    }
                }
            }
        }
    }
}

// MARK: - Breadcrumb

@Composable
private fun Breadcrumb(
    title: String,
    domainColor: Color,
    routes: List<PlaceholderRoute>,
    crumbs: List<PlaceholderCrumb>,
    showsDismiss: Boolean,
) {
    val scrollState = rememberScrollState()
    val path = crumbs.map { it.title } + title

    LaunchedEffect(path, scrollState.maxValue) {
        scrollState.scrollTo(scrollState.maxValue)
    }

    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        Row(
            modifier = Modifier
                .weight(1f)
                .horizontalScroll(scrollState),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(6.dp),
        ) {
            crumbs.forEach { crumb ->
                TitleMenu(crumb.title, crumb.routes, crumb.color)
                Text(
                    "›",
                    style = MaterialTheme.typography.titleMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
            TitleMenu(title, routes, domainColor)
        }
        if (showsDismiss) {
            DismissButton(domainColor)
        }
    }
}

@Composable
private fun TitleMenu(
    title: String,
    routes: List<PlaceholderRoute>,
    color: Color,
) {
    val visible = routes.filter { !it.isCurrent }
    var expanded by remember { mutableStateOf(false) }

    Box {
        Pill(
            title = title,
            color = color,
            dimmed = visible.isEmpty(),
            onClick = if (visible.isEmpty()) null else { -> expanded = true },
        )
        if (visible.isNotEmpty()) {
            DropdownMenu(expanded = expanded, onDismissRequest = { expanded = false }) {
                orderedRouteKinds(visible).forEach { kind ->
                    val kindRoutes = visible.filter { it.kind == kind }
                    if (kindRoutes.isNotEmpty()) {
                        Text(
                            kind.title,
                            style = MaterialTheme.typography.labelSmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                            modifier = Modifier.padding(horizontal = 12.dp, vertical = 4.dp),
                        )
                        kindRoutes.forEach { route ->
                            DropdownMenuItem(
                                text = { Text(route.label) },
                                onClick = {
                                    expanded = false
                                    route.action()
                                },
                            )
                        }
                    }
                }
            }
        }
    }
}

private fun orderedRouteKinds(routes: List<PlaceholderRoute>): List<PlaceholderRouteKind> =
    routes
        .map { it.kind }
        .distinct()
        .sortedWith(
            compareBy<PlaceholderRouteKind> { kind ->
                when (kind) {
                    PlaceholderRouteKind.Nav -> 0
                    is PlaceholderRouteKind.Modal -> 1
                    PlaceholderRouteKind.Child -> 2
                }
            }
                .thenBy { kind ->
                    when (kind) {
                        is PlaceholderRouteKind.Modal -> kind.style.index
                        else -> 0
                    }
                }
                .thenBy { it.title },
        )

@Composable
private fun Pill(
    title: String,
    color: Color,
    dimmed: Boolean,
    onClick: (() -> Unit)?,
) {
    Text(
        text = title,
        color = Color.Black,
        style = MaterialTheme.typography.bodyMedium,
        modifier = Modifier
            .alpha(if (dimmed) 0.5f else 1f)
            .clip(RoundedCornerShape(8.dp))
            .background(color)
            .then(if (onClick != null) Modifier.clickable { onClick() } else Modifier)
            .padding(horizontal = 8.dp, vertical = 3.dp),
    )
}

@Composable
private fun DismissButton(color: Color) {
    val dismiss = LocalPlaceholderDismiss.current
    Box(
        modifier = Modifier
            .clip(CircleShape)
            .background(color)
            .clickable { dismiss() }
            .padding(7.dp),
        contentAlignment = Alignment.Center,
    ) {
        Icon(
            Icons.Default.Close,
            contentDescription = "Dismiss",
            tint = Color.Black,
            modifier = Modifier.size(16.dp),
        )
    }
}
